src/bionics_ui.cpp

```diff
@@ -246,144 +277,113 @@ static void draw_connectors( const catacurses::window &win, const int start_y, c
-                                 TITLE_START_Y ) );
+    const int TITLE_WINDOW_Y = MAIN_WINDOW_Y + 1;
+    catacurses::window w_title = catacurses::newwin( TITLE_HEIGHT, WIDTH - 2, point( MAIN_WINDOW_X + 1,
+                                 TITLE_WINDOW_Y ) );
+
+    //Tabs bar
+    const int TAB_WINDOW_Y = TITLE_WINDOW_Y + TITLE_HEIGHT;
+    catacurses::window w_tabs = catacurses::newwin( TITLE_TAB_HEIGHT, WIDTH - 2,
+                                point( MAIN_WINDOW_X + 1,
+                                       TAB_WINDOW_Y ) );
+
+    const int header_line_y = TITLE_HEIGHT + TITLE_TAB_HEIGHT + 1;
+    const int footer_start_y = HEIGHT - FOOTER_HEIGHT - 1;
 
-    const int TAB_START_Y = TITLE_START_Y + 2;
-    //w_tabs is the tab bar for passive and active bionic groups
-    catacurses::window w_tabs = catacurses::newwin( TITLE_TAB_HEIGHT, WIDTH - 2, point( START_X + 1,
-                                TAB_START_Y ) );
+    const int col_right_bound = WIDTH - ( with_slots ? SLOTS_WIDTH : 3 );
 
     int scroll_position = 0;
     int cursor = 0;
+    int cur_tab_idx = 0;
 
-    //generate the tab title string and a count of the bionics owned
-    bionic_menu_mode menu_mode = ACTIVATING;
-    // offset for display: bionic with index i is drawn at y=list_start_y+i
-    // drawing the bionics starts with bionic[scroll_position]
-    // scroll_position;
-    const int list_start_y = HEADER_LINE_Y;
+    const int list_start_y = header_line_y + 1;
+    const int LIST_HEIGHT = footer_start_y - list_start_y - 1; //-1 is footer separator
     int half_list_view_location = LIST_HEIGHT / 2;
-    int max_scroll_position = std::max( 0, static_cast<int>( active.size() ) );
+    int max_scroll_position = std::max( 0, static_cast<int>( bionics_by_type[0].size() ) );
 
     input_context ctxt( "BIONICS" );
     ctxt.register_updown();
     ctxt.register_action( "ANY_INPUT" );
-    ctxt.register_action( "TOGGLE_EXAMINE" );
     ctxt.register_action( "REASSIGN" );
     ctxt.register_action( "NEXT_TAB" );
     ctxt.register_action( "PREV_TAB" );
     ctxt.register_action( "CONFIRM" );
     ctxt.register_action( "HELP_KEYBINDINGS" );
+    ctxt.register_action( "USAGE_HELP" );
 
     bool recalc = false;
     bool redraw = true;
 
     for( ;; ) {
         if( recalc ) {
-            passive = filtered_bionics( *my_bionics, TAB_PASSIVE );
-            active = filtered_bionics( *my_bionics, TAB_ACTIVE );
-
-            if( active.empty() && !passive.empty() ) {
-                tab_mode = TAB_PASSIVE;
+            bionics_by_type.clear();
+            for( auto b : *my_bionics ) {
+                bionics_displayType_id dType = b.id->display_type;
+                for( size_t i = 0; i < BionicsDisplayType::displayTypes.size(); i++ ) {
+                    if( BionicsDisplayType::displayTypes[i].ident() == dType ) {
+                        bionics_by_type[i].emplace_back( &b );
```

Comment:
Here you store the address of a temporary (only valid within this iteration of the loop).
Suggestion: change the loop to `auto &b`. Currently you iterate over copies of the elements in `my_bionics`.